File: cyberdeck.py
#!/usr/bin/env python3
"""
cyberdeck.py — Single transparent PySide6 window that sits as the desktop
background. Three columns (left: animations, center: graph+clock,
right: system shelf) + a bottom bar. Everything in ONE window — no
floating conky panels, no xwinwrap fights with i3.

Usage:  python3 cyberdeck.py [--screen WxH]
"""
import sys, os, signal, argparse, Xlib, Xlib.display, Xlib.Xatom

# Allow running from any directory
HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)

from PySide6.QtWidgets import QApplication, QWidget, QHBoxLayout, QVBoxLayout
from PySide6.QtCore    import Qt, QTimer
from PySide6.QtGui     import QPainter, QColor

from main_panel   import MainPanel
from right_panel  import RightPanel
from bottom_bar   import BottomBar
from spectrum_overlay import SpectrumOverlay


def setup_x11_below(wid):
    """
    Set window hints:
      - _NET_WM_STATE_BELOW + SKIP_TASKBAR + SKIP_PAGER
    """
    d = Xlib.display.Display()
    w = d.create_resource_object('window', wid)

    w.change_property(
        d.intern_atom('_NET_WM_STATE'),
        Xlib.Xatom.ATOM, 32,
        [
            d.intern_atom('_NET_WM_STATE_BELOW'),
            d.intern_atom('_NET_WM_STATE_SKIP_TASKBAR'),
            d.intern_atom('_NET_WM_STATE_SKIP_PAGER'),
        ]
    )

    d.sync()
    d.close()


def x11_focus_window(wid):
    """Explicitly set X input focus to a window (needed for override-redirect)."""
    try:
        import Xlib.X
        d = Xlib.display.Display()
        w = d.create_resource_object('window', wid)
        d.set_input_focus(w, Xlib.X.RevertToParent, Xlib.X.CurrentTime)
        d.sync()
        d.close()
    except Exception as e:
        print(f"[cyberdeck] focus error: {e}")


# Column width percentages
LEFT_PCT   = 28
CENTER_PCT = 44   # 100 - 28 - 28
RIGHT_PCT  = 28
BOTTOM_H   = 32   # px

# CYBERDECK_GLAVA=1 turns on the built-in, transparent bottom-up audio
# spectrum overlay across the graph area (left + center columns) — see
# spectrum_overlay.py. Off by default so the deck stays lightweight when
# you don't want it.
GLAVA_MODE = os.environ.get("CYBERDECK_GLAVA") == "1"


class CyberDeck(QWidget):
    """
    Single full-screen window on the desktop layer.
    i3 / the WM sees it as a desktop window and keeps it behind everything.
    Your normal work windows sit on top — you see the cyberdeck as the
    background when no window covers that zone.
    """
    def __init__(self, sw: int, sh: int):
        super().__init__()
        self.SW, self.SH = sw, sh

        # ── Window flags ──────────────────────────────────────────
        # BypassWindowManagerHint → override_redirect=True: i3 never tiles it.
        # Tool                    → no taskbar entry.
        # Input events are re-enabled after show via setup_x11_below which
        # calls XChangeWindowAttributes(event_mask=...) directly on the X11
        # window, so browser/FM widgets still receive keyboard and mouse.
        self.setWindowFlags(
            Qt.FramelessWindowHint |
            Qt.BypassWindowManagerHint |
            Qt.Tool
        )
        self.setAttribute(Qt.WA_TranslucentBackground)
        self.setAttribute(Qt.WA_ShowWithoutActivating)
        self.setGeometry(0, 0, sw, sh)
        self.setWindowTitle("cyberdeck")

        # Keep the window at the bottom of the X stacking order.
        # Override-redirect windows aren't managed by the WM, so we must
        # periodically lower ourselves to stay below all normal windows.
        self._lower_timer = QTimer(self)
        self._lower_timer.timeout.connect(self._lower_below)
        self._lower_timer.start(5000)  # every 5s — only needed if WM pushes us up

        # ── Layout ────────────────────────────────────────────────
        root = QVBoxLayout(self)
        root.setContentsMargins(0, 0, 0, 0)
        root.setSpacing(0)

        # Top strip: 3 columns
        cols_widget = QWidget(self)
        cols_widget.setAttribute(Qt.WA_TranslucentBackground)
        cols = QHBoxLayout(cols_widget)
        cols.setContentsMargins(0, 0, 0, 0)
        cols.setSpacing(0)

        lw = sw * LEFT_PCT   // 100
        rw = sw * RIGHT_PCT  // 100
        cw = sw - lw - rw

        top_h = sh - BOTTOM_H

        # Left + center are now a single merged panel — one pixmap, one
        # paintEvent, no crop_x seam between two separately-drawn widgets.
        self._main  = MainPanel(cols_widget)
        self._right = RightPanel(cols_widget)

        self._main .setFixedSize(lw + cw, top_h)
        self._right.setFixedSize(rw, top_h)

        cols.addWidget(self._main)
        cols.addWidget(self._right)

        # ── Bottom-up audio spectrum, overlaid on the graph area ────────
        # Sits above the left+center columns (same combined width/height
        # as the graph they share), transparent and click-through, so the
        # graph and pipes texture stay visible underneath it.
        self._spectrum = None
        if GLAVA_MODE:
            self._spectrum = SpectrumOverlay(self)
            self._spectrum.setGeometry(0, 0, lw + cw, top_h)

        # Bottom bar: a separate top-level DOCK window. i3 manages docks
        # specially — always visible, screen space reserved (like i3bar).
        self._bottom = BottomBar()
        self._bottom.setWindowFlags(Qt.FramelessWindowHint)
        self._bottom.setAttribute(Qt.WA_X11NetWmWindowTypeDock)
        self._bottom.setAttribute(Qt.WA_ShowWithoutActivating)
        self._bottom.setFixedSize(sw, BOTTOM_H)
        self._bottom.move(0, sh - BOTTOM_H)
        self._bottom.setWindowTitle("sworddeck-bar")

        root.addWidget(cols_widget)
        root.addStretch(1)

        if self._spectrum is not None:
            self._spectrum.raise_()

        # ── Graph render triggered only by graph-edit.sh (no polling) ──

    def _lower_below(self):
        """Keep cyberdeck below normal windows.

        Only calls lower() when the window is NOT already at the bottom of the
        X11 stack — avoids the constant ConfigureNotify → repaint cycle that
        caused visible flicker every second.
        """
        try:
            import Xlib.display as _xd
            d  = _xd.Display()
            wm = d.screen().root.query_tree().children
            our_wid = int(self.winId())
            # wm children are ordered bottom→top; we want to be first (index 0)
            # If we are already at or near the bottom, skip lower()
            if wm and int(wm[0].id) == our_wid:
                d.close()
                return
            d.close()
        except Exception:
            pass
        self.lower()

    def paintEvent(self, _):
        # Solid dark fill behind everything (panels + spectrum overlay sit
        # on top of this same flat background).
        p = QPainter(self)
        p.fillRect(self.rect(), QColor(40, 44, 52, 255))
        p.end()

    def keyPressEvent(self, e):
        # Ctrl+C / Escape quit
        if e.key() in (Qt.Key_Escape,) and e.modifiers() & Qt.ControlModifier:
            QApplication.quit()

    def showEvent(self, e):
        super().showEvent(e)
        QTimer.singleShot(300, self._apply_x11)

    def _apply_x11(self):
        wid = int(self.winId())
        try:
            setup_x11_below(wid)
            # Give MainPanel the WID so it can grab X focus when switching
            # to interactive tabs (browser, FM, terminal)
            self._main.set_deck_wid(wid)
            print(f"[cyberdeck] X11 BELOW hints applied (window {wid})")
        except Exception as e:
            print(f"[cyberdeck] X11 error: {e}")


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--screen", default="1920x1080",
                    help="WxH e.g. 1920x1080")
    args = ap.parse_args()
    try:
        sw, sh = map(int, args.screen.split("x"))
    except Exception:
        sw, sh = 1920, 1080

    # Handle SIGTERM cleanly (so cyberdesk.sh stop works)
    signal.signal(signal.SIGTERM, lambda *_: QApplication.quit())

    app = QApplication(sys.argv)
    app.setApplicationName("cyberdeck")

    # Auto-detect screen size if not overridden
    screen = app.primaryScreen()
    if args.screen == "1920x1080":
        sw = screen.size().width()
        sh = screen.size().height()

    win = CyberDeck(sw, sh)
    win.show()
    win._bottom.show()

    sys.exit(app.exec())


if __name__ == "__main__":
    main()

File: cpp-sworddeck/src/cyberdeck.cpp
#include "cyberdeck.h"
#include "main_panel.h"
#include "right_panel.h"
#include "bottom_bar.h"
#include "spectrum_overlay.h"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QPainter>
#include <QTimer>
#include <QKeyEvent>
#include <QApplication>
#include <QScreen>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <cstdlib>
#include <string>

static bool s_glavaMode = false;

CyberDeck::CyberDeck(int sw, int sh, QWidget *parent)
    : QWidget(parent), m_sw(sw), m_sh(sh),
      m_main(nullptr), m_right(nullptr), m_bottom(nullptr), m_spectrum(nullptr)
{
    if (std::getenv("CYBERDECK_GLAVA"))
        s_glavaMode = (std::string(std::getenv("CYBERDECK_GLAVA")) == "1");

    setWindowFlags(Qt::FramelessWindowHint | Qt::BypassWindowManagerHint | Qt::Tool);
    setAttribute(Qt::WA_TranslucentBackground);
    setAttribute(Qt::WA_ShowWithoutActivating);
    setGeometry(0, 0, sw, sh);
    setWindowTitle("cyberdeck");

    /* Periodically force the cyberdeck below all windows.
     * BypassWindowManagerHint prevents the WM from managing us,
     * but we still need to actively lower ourselves below
     * normal client windows that get raised. */
    auto *lowerTimer = new QTimer(this);
    connect(lowerTimer, &QTimer::timeout, this, [this]() {
        if (isVisible() && isActiveWindow()) lower();
    });
    lowerTimer->start(5000);

    auto *root = new QVBoxLayout(this);
    root->setContentsMargins(0, 0, 0, 0);
    root->setSpacing(0);

    auto *colsWidget = new QWidget(this);
    colsWidget->setAttribute(Qt::WA_TranslucentBackground);
    auto *cols = new QHBoxLayout(colsWidget);
    cols->setContentsMargins(0, 0, 0, 0);
    cols->setSpacing(0);

    int lw = sw * LEFT_PCT / 100;
    int rw = sw * RIGHT_PCT / 100;
    int cw = sw - lw - rw;
    int topH = sh - BOTTOM_H;

    m_main = new MainPanel(colsWidget);
    m_right = new RightPanel(colsWidget);
    m_main->setFixedSize(lw + cw, topH);
    m_right->setFixedSize(rw, topH);
    cols->addWidget(m_main);
    cols->addWidget(m_right);

    if (s_glavaMode) {
        m_spectrum = new SpectrumOverlay(this);
        m_spectrum->setGeometry(0, 0, lw + cw, topH);
    }

    m_bottom = new BottomBar(this);
    m_bottom->setWindowFlags(Qt::Window | Qt::FramelessWindowHint | Qt::BypassWindowManagerHint);
    m_bottom->setAttribute(Qt::WA_X11NetWmWindowTypeDock);
    m_bottom->setAttribute(Qt::WA_ShowWithoutActivating);
    m_bottom->setFixedSize(sw, BOTTOM_H);
    m_bottom->move(0, sh - BOTTOM_H);
    m_bottom->setWindowTitle("sworddeck-bar");

    root->addWidget(colsWidget);
    root->addStretch(1);

    if (m_spectrum) m_spectrum->raise();
}

void CyberDeck::showBottomBar() {
    if (m_bottom) {
        m_bottom->show();
        /* Apply hints to bottom bar once it has a window ID */
        QTimer::singleShot(300, this, [this]() {
            if (m_bottom) setupX11Hints(m_bottom->winId(), /*isDock=*/true);
        });
    }
}

void CyberDeck::lowerBelow() {
    /* Intentionally empty — kept so the timer slot signature is valid.
     * Timer is no longer started; method retained for ABI compatibility. */
}

void CyberDeck::paintEvent(QPaintEvent *) {
    QPainter p(this);
    p.fillRect(rect(), QColor(40, 44, 52, 255));
}

void CyberDeck::keyPressEvent(QKeyEvent *e) {
    if (e->key() == Qt::Key_Escape && e->modifiers() & Qt::ControlModifier)
        QApplication::quit();
}

void CyberDeck::showEvent(QShowEvent *e) {
    QWidget::showEvent(e);
    QTimer::singleShot(300, this, &CyberDeck::applyX11Hints);
    QTimer::singleShot(500, this, [this]() { lower(); });
}

void CyberDeck::applyX11Hints() {
    setupX11Hints(winId(), /*isDock=*/false);
}

void CyberDeck::setupX11Hints(WId wid, bool isDock) {
    Display *d = XOpenDisplay(nullptr);
    if (!d) return;

    Atom wmState = XInternAtom(d, "_NET_WM_STATE", False);
    Atom states[3] = {
        XInternAtom(d, "_NET_WM_STATE_BELOW",        False),
        XInternAtom(d, "_NET_WM_STATE_SKIP_TASKBAR", False),
        XInternAtom(d, "_NET_WM_STATE_SKIP_PAGER",   False)
    };
    XChangeProperty(d, wid, wmState, XA_ATOM, 32, PropModeReplace,
                    (unsigned char*)states, 3);

    if (isDock) {
        /* Dock window type so panels/bars reserve strut space */
        Atom wmType  = XInternAtom(d, "_NET_WM_WINDOW_TYPE",      False);
        Atom dockType= XInternAtom(d, "_NET_WM_WINDOW_TYPE_DOCK", False);
        XChangeProperty(d, wid, wmType, XA_ATOM, 32, PropModeReplace,
                        (unsigned char*)&dockType, 1);
    }

    XFlush(d);
    XCloseDisplay(d);
    qInfo("[cyberdeck] X11 hints applied to window %lu (dock=%d)", wid, isDock);
}
